lib/logstash/outputs/elasticsearch/common.rb

```diff
@@ -0,0 +1,171 @@
+                       :message => e.message,
+                       :class => e.class.name,
+                       :backtrace => e.backtrace)
+        end
+        retries_left -= 1
+      end
+    end
+
+    def submit(actions)
+      es_actions = actions.map { |a, doc, event| [a, doc, event.to_hash] }
+
+      bulk_response = safe_bulk(es_actions,actions)
+
+      # If there are no errors, we're done here!
+      return unless bulk_response["errors"]
+
+      actions_to_retry = []
+      bulk_response["items"].each_with_index do |response,idx|
+        action_type, action_props = response.first
+        status = action_props["status"]
+        action = actions[idx]
+
+        if SUCCESS_CODES.include?(status)
+          next
+        elsif RETRYABLE_CODES.include?(status)
+          @logger.warn "retrying failed action with response code: #{status}"
+          actions_to_retry << action
+        else
+          @logger.warn "Failed action. ", status: status, action: action, response: response
+        end
+      end
+
+      actions_to_retry
+    end
+
+    # get the action parameters for the given event
+    def event_action_params(event)
+      type = get_event_type(event)
+
+      params = {
+        :_id => @document_id ? event.sprintf(@document_id) : nil,
+        :_index => event.sprintf(@index),
+        :_type => type,
+        :_routing => @routing ? event.sprintf(@routing) : nil
+      }
+
+      params[:_upsert] = LogStash::Json.load(event.sprintf(@upsert)) if @action == 'update' && @upsert != ""
+      params
+    end
+
+    # Determine the correct value for the 'type' field for the given event
+    def get_event_type(event)
+      # Set the 'type' value for the index.
+      type = if @document_type
+               event.sprintf(@document_type)
+             else
+               event["type"] || "logs"
+             end
+
+      if !(type.is_a?(String) || type.is_a?(Numeric))
+        @logger.warn("Bad event type! Non-string/integer type value set!", :type_class => type.class, :type_value => type.to_s, :event => event)
+      end
+
+      type.to_s
+    end
+
+    # Rescue retryable errors during bulk submission
+    def safe_bulk(es_actions,actions)
+      @client.bulk(es_actions)
+    rescue Manticore::SocketException, Manticore::SocketTimeout => e
+      # If we can't even connect to the server let's just print out the URL (:hosts is actually a URL)
+      # and let the user sort it out from there
+      @logger.error(
+        "Attempted to send a bulk request to Elasticsearch configured at '#{@client.client_options[:hosts]}',"+
+          " but Elasticsearch appears to be unreachable or down!",
+        :client_config => @client.client_options,
+        :error_message => e.message,
+        :class => e.class.name
+      )
+      @logger.debug("Failed actions for last bad bulk request!", :actions => actions)
+
+      # We retry until there are no errors! Errors should all go to the retry queue
+      sleep 1
```

Comment:
let me know if this crosses the line in the form of a scope-creep.... adding exponential backoff to the retry intervals.